dallinger/experiment_server/worker_events.py

```diff
@@ -109,66 +149,71 @@ def worker_function(
-        participants = models.Participant.query.filter_by(
-            assignment_id=assignment_id
-        ).all()
-
-        # if there are one or more participants select the most recent
-        if participants:
-            participant = max(participants, key=attrgetter("creation_time"))
-
-        # if there are none print an error
-        else:
+        if not participant:
             exp.log(
                 "Warning: No participants associated with this "
                 "assignment_id. Notification will not be processed.",
                 key,
             )
-            return None
-
-    elif participant_id is not None:
-        # XXX Why is this not a Participant.get()?
-        participant = models.Participant.query.filter_by(id=participant_id).all()[0]
-    else:
+            return
+    elif not participant and not node:
         raise ValueError(
             "Error: worker_function needs either an assignment_id or a "
             "participant_id, they cannot both be None"
         )
 
+    # Distinguish between the time of the event and the time it was pulled off
+    # the queue for processing
     runner = runner_cls(
-        participant, assignment_id, exp, db.session, _config(), datetime.now()
+        participant,
+        assignment_id,
+        exp,
+        db.session,
+        config=_config(),
+        now=datetime.now(),
+        receive_time=receive_time,
+        node=node,
+        details=details,
     )
     runner()
     db.session.commit()
 
 
-class WorkerEvent(object):
-    key = "-----"
+class _WorkerMeta(type):
+    _WORKER_EVENTS = {}
 
-    supported_event_types = (
-        "AssignmentAccepted",
-        "AssignmentAbandoned",
-        "AssignmentReassigned",
-        "AssignmentReturned",
-        "AssignmentSubmitted",
-        "BotAssignmentSubmitted",
-        "BotAssignmentRejected",
-        "NotificationMissing",
-    )
+    def __init__(cls, name, bases, dct):
+        """Register subclasses with a name registry"""
+        cls._WORKER_EVENTS[name] = cls
 
-    @classmethod
     def for_name(cls, name):
-        if name in cls.supported_event_types:
-            return globals()[name]
+        return cls._WORKER_EVENTS.get(name)
+
 
-    def __init__(self, participant, assignment_id, experiment, session, config, now):
+class WorkerEvent(metaclass=_WorkerMeta):
+    key = "-----"
+
+    def __init__(
+        self,
+        participant=None,
+        assignment_id=None,
+        experiment=None,
+        session=None,
+        config=None,
+        now=None,
+        receive_time=None,
+        node=None,
+        details=None,
+    ):
         self.participant = participant
         self.assignment_id = assignment_id
         self.experiment = experiment
         self.session = session
         self.config = config
         self.now = now
+        self.receive_time = receive_time
+        self.node = node
+        self.details = details
 
     @property
     def data(self):
```

Comment:
Can we add `receive_time` and `details` to the `data` property? (`details` is just data, right?)

Reply:
@jessesnyder I don't see why not, they'd still be instance attributes, right? What is the `data` property used for?

Reply:
Generally `details` should be JSON-ish data that could get stored into the `details` JSONB field on an `Event` object in the DB (as it is already done in the `TrackingEvent` special case), but it could be used for whatever by one of these custom event handlers.

Reply:
I've added those attributes to the data property.